runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillConnection.java

```diff
@@ -48,17 +49,21 @@ public static Builder builder() {
 
   public abstract String backendWorkerToken();
 
-  public abstract Optional<WindmillServiceAddress> directEndpoint();
+  abstract Optional<WindmillServiceAddress> directEndpoint();
 
-  public abstract CloudWindmillServiceV1Alpha1Stub stub();
+  abstract Supplier<CloudWindmillServiceV1Alpha1Stub> stubFactory();
+
+  public final CloudWindmillServiceV1Alpha1Stub newStub() {
```

Comment:
if it's a supplier, I think currentStub() or something could be better name since new implies it is freshly allocated and not reused